processing/src/main/java/org/apache/druid/segment/serde/ComplexMetricSerde.java

```diff
@@ -142,14 +120,100 @@ public Object fromBytes(byte[] data, int start, int numBytes)
+   */
+  public void deserializeColumn(
+      ByteBuffer buffer,
+      ColumnBuilder builder,
+      ColumnConfig columnConfig
+  )
+  {
+    deserializeColumn(buffer, builder);
+  }
+
+
+  /**
+   * {@link ComplexMetricSerde#deserializeColumn(ByteBuffer, ColumnBuilder, ColumnConfig)} should be used instead of this.
+   * This method is left for backward compatibility.
    */
+  @Deprecated
+  public void deserializeColumn(ByteBuffer buffer, ColumnBuilder builder)
+  {
+    // default implementation to match default serializer implementation
+    final int position = buffer.position();
+    final byte version = buffer.get();
+    if (version == CompressedComplexColumnSerializer.IS_COMPRESSED) {
+      CompressedComplexColumnSupplier supplier = CompressedComplexColumnSupplier.read(
+          buffer,
+          builder,
+          getTypeName(),
+          getObjectStrategy()
+      );
+      builder.setComplexColumnSupplier(supplier);
+      builder.setNullValueIndexSupplier(supplier.getNullValues());
+      builder.setHasNulls(!supplier.getNullValues().isEmpty());
+    } else {
+      buffer.position(position);
+      builder.setComplexColumnSupplier(
+          new ComplexColumnPartSupplier(
+              getTypeName(),
+              GenericIndexed.read(buffer, getObjectStrategy(), builder.getFileMapper())
+          )
+      );
+    }
+  }
+
+  /**
+   * {@link ComplexMetricSerde#getSerializer(SegmentWriteOutMedium, String, IndexSpec)} should be used instead of this.
+   * This method is left for backward compatibility.
+   */
+  @Nullable
+  @Deprecated
   public GenericColumnSerializer getSerializer(SegmentWriteOutMedium segmentWriteOutMedium, String column)
   {
-    return ComplexColumnSerializer.create(segmentWriteOutMedium, column, this.getObjectStrategy());
+    return null;
+  }
+
+  /**
+   * This method provides the ability for a ComplexMetricSerde to control its own serialization.
+   * Default implementation uses {@link CompressedComplexColumnSerializer} if {@link IndexSpec#complexMetricCompression}
+   * is not null or uncompressed/none, or {@link LargeColumnSupportedComplexColumnSerializer} if no compression is
+   * specified.
+   *
+   * @return an instance of {@link GenericColumnSerializer} used for serialization.
+   */
+  public GenericColumnSerializer getSerializer(
+      SegmentWriteOutMedium segmentWriteOutMedium,
+      String column,
+      IndexSpec indexSpec
+  )
+  {
+    // backwards compatibility, if defined use it
+    final GenericColumnSerializer serializer = getSerializer(segmentWriteOutMedium, column);
+    if (serializer != null) {
+      return serializer;
+    }
+
+    // otherwise, use compressed or generic indexed based serializer
+    CompressionStrategy strategy = indexSpec.getComplexMetricCompression();
+    if (strategy == null || CompressionStrategy.NONE == strategy || CompressionStrategy.UNCOMPRESSED == strategy) {
+      return LargeColumnSupportedComplexColumnSerializer.create(
```

Comment:
`LargeColumnSupportedComplexColumnSerializer` is basically identical to `ComplexColumnSerializer`, they both use `GenericIndexedWriter` with a filename of `StringUtils.format("%s.complex_column", filenameBase)`, the main difference is that  `LargeColumnSupportedComplexColumnSerializer` passes the `FileSmoosher` through to the writer allowing it to write v2 `GenericIndexed` which has the multi file support. So I think there shouldn't be any compatibility concerns because on the read side since both of these just use `GenericIndexed.read`